File: PoseKit/Tests/PoseKitTests/LimbTests.swift
import XCTest
@testable import PoseKit

final class LimbTests: XCTestCase {
    let standing = Fixtures.standing

    func testIdenticalPoseAllBonesScoreOne() {
        let scores = LimbSimilarity.boneScores(reference: standing, live: standing)
        XCTAssertEqual(scores.count, 10)
        for (_, s) in scores { XCTAssertEqual(s, 1.0, accuracy: 1e-5) }
    }

    func testOppositeLimbScoresZero() {
        var p = standing.points
        // Point left forearm straight up instead of down (opposite direction).
        let elbow = p[.leftElbow]!
        let wrist = p[.leftWrist]!
        p[.leftWrist] = elbow - (wrist - elbow)
        let live = PoseVector(points: p)
        let scores = LimbSimilarity.boneScores(reference: standing, live: live)
        XCTAssertEqual(scores[.leftForearm]!, 0.0, accuracy: 1e-5)
        XCTAssertEqual(scores[.rightForearm]!, 1.0, accuracy: 1e-5)
    }

    func testPerpendicularLimbScoresHalf() {
        var p = standing.points
        let elbow = p[.leftElbow]!
        let wrist = p[.leftWrist]!
        let d = wrist - elbow
        p[.leftWrist] = elbow + SIMD2<Float>(-d.y, d.x) // rotate bone 90°
        let scores = LimbSimilarity.boneScores(reference: standing, live: PoseVector(points: p))
        XCTAssertEqual(scores[.leftForearm]!, 0.5, accuracy: 1e-5)
    }

    func testMissingEndpointOmitsBone() {
        var p = standing.points
        p[.leftWrist] = nil
        let scores = LimbSimilarity.boneScores(reference: standing, live: PoseVector(points: p))
        XCTAssertNil(scores[.leftForearm])
        XCTAssertEqual(scores.count, 9)
    }

    func testWorstBoneIdentified() {
        var p = standing.points
        let elbow = p[.rightElbow]!
        let wrist = p[.rightWrist]!
        p[.rightWrist] = elbow - (wrist - elbow)
        let worst = LimbSimilarity.worstBone(reference: standing, live: PoseVector(points: p))
        XCTAssertEqual(worst?.bone, .rightForearm)
    }

    func testMeanScoreNilWhenNoBones() {
        let empty = PoseVector(points: [.nose: [0.5, 0.1]])
        XCTAssertNil(LimbSimilarity.meanScore(reference: standing, live: empty))
    }
}

File: android/posekit/src/test/kotlin/com/oerol/posekit/LimbTests.kt
package com.oerol.posekit

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class LimbTests {
    private val standing = Fixtures.standing

    @Test
    fun identicalPoseAllBonesScoreOne() {
        val scores = LimbSimilarity.boneScores(standing, standing)
        assertEquals(10, scores.size)
        for ((_, s) in scores) assertEquals(1.0f, s, 1e-5f)
    }

    @Test
    fun oppositeLimbScoresZero() {
        val p = standing.points.toMutableMap()
        // Point left forearm straight up instead of down (opposite direction).
        val elbow = p.getValue(Joint.leftElbow)
        val wrist = p.getValue(Joint.leftWrist)
        p[Joint.leftWrist] = elbow - (wrist - elbow)
        val live = PoseVector(p)
        val scores = LimbSimilarity.boneScores(standing, live)
        assertEquals(0.0f, scores.getValue(Bone.LEFT_FOREARM), 1e-5f)
        assertEquals(1.0f, scores.getValue(Bone.RIGHT_FOREARM), 1e-5f)
    }

    @Test
    fun perpendicularLimbScoresHalf() {
        val p = standing.points.toMutableMap()
        val elbow = p.getValue(Joint.leftElbow)
        val wrist = p.getValue(Joint.leftWrist)
        val d = wrist - elbow
        p[Joint.leftWrist] = elbow + Vec2(-d.y, d.x) // rotate bone 90°
        val scores = LimbSimilarity.boneScores(standing, PoseVector(p))
        assertEquals(0.5f, scores.getValue(Bone.LEFT_FOREARM), 1e-5f)
    }

    @Test
    fun missingEndpointOmitsBone() {
        val p = standing.points.toMutableMap()
        p.remove(Joint.leftWrist)
        val scores = LimbSimilarity.boneScores(standing, PoseVector(p))
        assertNull(scores[Bone.LEFT_FOREARM])
        assertEquals(9, scores.size)
    }

    @Test
    fun worstBoneIdentified() {
        val p = standing.points.toMutableMap()
        val elbow = p.getValue(Joint.rightElbow)
        val wrist = p.getValue(Joint.rightWrist)
        p[Joint.rightWrist] = elbow - (wrist - elbow)
        val worst = LimbSimilarity.worstBone(standing, PoseVector(p))
        assertEquals(Bone.RIGHT_FOREARM, worst?.first)
    }

    @Test
    fun meanScoreNullWhenNoBones() {
        val empty = PoseVector(mapOf(Joint.nose to Vec2(0.5f, 0.1f)))
        assertNull(LimbSimilarity.meanScore(standing, empty))
    }
}
